### PR TITLE
Bucket reader: Initialize new query stats struct at each goroutine

### PR DESCRIPTION
**Pull request by yeya24**
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Instead of using a single lock and use the lock to protect access to query stats field, initialize a new query stats at the beginning of each goroutine, and finally merge the stats into the reader's stats. Then each goroutine only needs to hold lock once at the end of its goroutine.

## Verification

Existing query stats test should still pass.

### CODE REVIEW

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -3159,10 +3161,11 @@ func (r *bucketIndexReader) loadSeries(ctx context.Context, ids []storage.Series
 			return r.loadSeries(ctx, ids[i:], true, uint64(id), uint64(id)+uint64(n+int(l)+1), bytesLimiter, tenant)
 		}
 		c = c[n : n+int(l)]
-		r.mtx.Lock()
+		r.loadedSeriesMtx.Lock()
 		r.loadedSeries[id] = c
+		r.loadedSeriesMtx.Unlock()
+
 		r.block.indexCache.StoreSeries(r.block.meta.ULID, id, c, tenant)
-		r.mtx.Unlock()
```
Comment (yeya24) [inline pkg/store/bucket.go:3165]: This is a major change.
Previously the code was like below. `StoreSeries` can be time consuming and doesn't need to be protected by the lock.

```
r.mtx.Lock()
r.loadedSeries[id] = c
r.block.indexCache.StoreSeries(r.block.meta.ULID, id, c, tenant)
r.mtx.Unlock()
```

File: pkg/store/bucket.go [hunk 2]
```diff
@@ -2928,6 +2930,12 @@ func (r *bucketIndexReader) fetchPostings(ctx context.Context, keys []labels.Lab
 		// Fetch from object storage concurrently and update stats and posting list.
 		g.Go(func() error {
 			begin := time.Now()
+			stats := new(queryStats)
+			defer func() {
+				r.mtx.Lock()
+				r.stats = r.stats.merge(stats)
```
Comment (GiedriusS) [inline pkg/store/bucket.go:2936]: Maybe we could just move this usage of `r.mtx` inside of `r.stats.merge` i.e. `r.stats.merge()` could use a separate mutex underneath? Then we would avoid having to do this dance the whole time.

File: pkg/store/bucket.go [hunk 3]
```diff
@@ -2382,8 +2382,10 @@ type bucketIndexReader struct {
 	dec   *index.Decoder
 	stats *queryStats
 
-	mtx          sync.Mutex
-	loadedSeries map[storage.SeriesRef][]byte
+	mtx sync.Mutex
+	// Separate mutex to protect loadedSeries.
```
Comment (GiedriusS) [inline pkg/store/bucket.go:2386]: I think it's obvious, I suggest removing this comment.

File: pkg/store/bucket.go [hunk 4]
```diff
@@ -3576,6 +3570,8 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesEntry, a
 // save saves a copy of b's payload to a memory pool of its own and returns a new byte slice referencing said copy.
 // Returned slice becomes invalid once r.block.chunkPool.Put() is called.
 func (r *bucketChunkReader) save(b []byte) ([]byte, error) {
+	r.mtx.Lock()
```
Comment (GiedriusS) [inline pkg/store/bucket.go:3573]: Then we can rename `r.mtx` to `r.chunkBytesMtx`.

File: pkg/store/bucket.go (hunk 3 above)
Comment (yeya24) [reply pkg/store/bucket.go:2386]: Made the change.

File: pkg/store/bucket.go (hunk 2 above)
Comment (yeya24) [reply pkg/store/bucket.go:2936]:
```
func (s queryStats) merge(o *queryStats) *queryStats
```

This is the current `merge` function and it takes the value as receiver (not pointer). This seems not that easy to add a lock to `queryStats` struct itself. WDYT?
Is it fine to keep the current code

File: pkg/store/bucket.go (hunk 4 above)
Comment (yeya24) [reply pkg/store/bucket.go:3573]: If we don't embed a lock to `queryStats` I will keep the mutex here

File: pkg/store/bucket.go (hunk 2 above)
Comment (GiedriusS) [reply pkg/store/bucket.go:2935]: Why not convert it into `func (s *queryStats) merge(o *queryStats) *queryStats`? :thinking:

Comment (yeya24) [reply pkg/store/bucket.go:2935]: I thought the idea is to return a new queryStats object every time. In this case, we will return the same one. But I guess it should be ok, let me try.

**Comment by yeya24** [comment]
Should be ready for another round of review.

**Review by GiedriusS — APPROVED** [review]
:+1: nice cleanup

**Pull request merged**